src/storage/exec/ScanNode.h

```diff
@@ -0,0 +1,293 @@
+                              bool enableReadFollower,
+                              int64_t limit,
+                              std::unordered_map<PartitionID, cpp2::ScanCursor>* cursors,
+                              nebula::DataSet* resultDataSet)
+      : context_(context),
+        tagNodes_(std::move(tagNodes)),
+        enableReadFollower_(enableReadFollower),
+        limit_(limit),
+        cursors_(cursors),
+        resultDataSet_(resultDataSet) {
+    name_ = "ScanVertexPropNode";
+    for (std::size_t i = 0; i < tagNodes_.size(); ++i) {
+      tagNodesIndex_.emplace(tagNodes_[i]->tagId(), i);
+    }
+  }
+
+  nebula::cpp2::ErrorCode doExecute(PartitionID partId, const Cursor& cursor) override {
+    auto ret = RelNode::doExecute(partId);
+    if (ret != nebula::cpp2::ErrorCode::SUCCEEDED) {
+      return ret;
+    }
+
+    std::string start;
+    std::string prefix = NebulaKeyUtils::vertexPrefix(partId);
+    if (cursor.empty()) {
+      start = prefix;
+    } else {
+      start = cursor;
+    }
+
+    std::unique_ptr<kvstore::KVIterator> iter;
+    auto kvRet = context_->env()->kvstore_->rangeWithPrefix(
+        context_->planContext_->spaceId_, partId, start, prefix, &iter, enableReadFollower_);
+    if (kvRet != nebula::cpp2::ErrorCode::SUCCEEDED) {
+      return kvRet;
+    }
+
+    const auto rowLimit = limit_;
+    auto vIdLen = context_->vIdLen();
+    auto isIntId = context_->isIntId();
+    std::string currentVertexId;
+    for (; iter->valid(); iter->next()) {
+      auto key = iter->key();
+      auto tagId = NebulaKeyUtils::getTagId(vIdLen, key);
+      auto tagIdIndex = tagNodesIndex_.find(tagId);
+      if (tagIdIndex != tagNodesIndex_.end()) {
+        break;
+      }
+    }
+    if (iter->valid()) {
+      auto vIdSlice = NebulaKeyUtils::getVertexId(vIdLen, iter->key());
+      currentVertexId = vIdSlice.subpiece(0, vIdSlice.find_first_of('\0')).toString();
+    }
+    for (; iter->valid() && static_cast<int64_t>(resultDataSet_->rowSize()) < rowLimit;
+         iter->next()) {
+      auto key = iter->key();
+      auto tagId = NebulaKeyUtils::getTagId(vIdLen, key);
+      auto tagIdIndex = tagNodesIndex_.find(tagId);
+      if (tagIdIndex == tagNodesIndex_.end()) {
+        continue;
+      }
+      auto vIdSlice = NebulaKeyUtils::getVertexId(vIdLen, key);
+      auto vertexId = vIdSlice.subpiece(0, vIdSlice.find_first_of('\0'));
+      if (vertexId != currentVertexId) {
+        collectOneRow(isIntId, vIdLen, currentVertexId);
+        currentVertexId = vertexId;
+      }  // collect vertex row
+      if (static_cast<int64_t>(resultDataSet_->rowSize()) >= rowLimit) {
+        break;
+      }
+      auto value = iter->val();
+      tagNodes_[tagIdIndex->second]->doExecute(key.toString(), value.toString());
+    }  // iterate key
+    if (static_cast<int64_t>(resultDataSet_->rowSize()) < rowLimit) {
+      collectOneRow(isIntId, vIdLen, currentVertexId);
+    } else {
+      for (auto& tagNode : tagNodes_) {
```

Comment:
The clear here seems useless? Because `collectOneRow` has invoke the function. We need to clear the key/value in `TagNode` once we have called `collectOneRow`.

Reply:
Yes, we had avoid it in previous code.